tests/regression/src/lib.rs

```diff
@@ -87,70 +93,202 @@ mod tests {
+        create_dir_all(Path::new(&directory)).unwrap();
+        let annotate_file = format!("target/perf_outputs/{test_name}_{commit_hash}.annotated.txt");
+        let mut file = File::create(annotate_file).expect("Failed to create annotation file");
+        file.write_all(&output.stdout)
             .expect("Failed to write annotation file");
+    }
 
-        let count = find_instruction_count(&annotate_file)
-            .expect("Failed to get instruction count from file");
-        // This is temporary code to showcase the future diff functionality, here the code regresses by 10% each time so this test will almost always fail
-        let new_count = count + count / 10;
-        let diff = new_count - count;
-        assert!(diff <= self::MAX_DIFF, "Instruction count difference in {} exceeds the threshold, regression of {} instructions", test_name, diff);
+    /// Function to run the diff test using valgrind, only called when diff mode is set
+    fn run_diff_test(test_name: &str) {
+        let (prev_file, curr_file) = find_and_validate_diff_files(test_name);
+        print!("{prev_file} and {curr_file}");
+        let diff_output = run_cg_annotate_diff(&prev_file, &curr_file);
+        save_diff_output(&diff_output, test_name);
+
+        let diff = find_instruction_count(&diff_output)
+            .expect("Failed to parse cg_annotate --diff output");
+
+        assert_diff_within_threshold(diff, test_name);
+    }
+
+    fn find_and_validate_diff_files(test_name: &str) -> (String, String) {
+        let annotated_files = find_annotated_files(test_name);
+        let file_len = annotated_files.len();
+        if file_len != 2 {
+            panic!("Expected exactly 2 annotated files for {test_name}, found {file_len}");
+        }
+
+        let file1 = &annotated_files[0];
+        let file2 = &annotated_files[1];
+
+        if !is_commit_in_log(file1) || !is_commit_in_log(file2) {
+            let first_hash = extract_commit_hash(file1);
+            let second_hash = extract_commit_hash(file2);
+            panic!(
+                "One or both commit hashes are not in the git log: {first_hash} or {second_hash}"
+            );
+        }
+
+        let (old_file, new_file) = if is_older_commit(file1, file2) {
+            (file1.clone(), file2.clone())
+        } else if is_older_commit(file2, file1) {
+            (file2.clone(), file1.clone())
+        } else {
+            panic!("Cannot determine the older commit between {file1} and {file2}",);
+        };
+
+        (old_file, new_file)
+    }
+
+    fn find_annotated_files(test_name: &str) -> Vec<String> {
+        let pattern = format!("target/cg_artifacts/cachegrind_{test_name}_*.out");
+        glob::glob(&pattern)
+            .expect("Failed to read glob pattern")
+            .filter_map(Result::ok)
+            .map(|path| path.to_string_lossy().into_owned())
+            .collect()
+    }
+
+    fn is_commit_in_log(file: &str) -> bool {
```

Comment:
I don't think they necessarily need to be in a different file, but I would vote for moving `is_commit_in_log`, `is_older_commit`, and `current_commit` to a separate `git` module.

```
mod git {
   fn get_current_commit() -> String;
}

git::get_current_commit()
```